### PR TITLE
feat: add memoization to when converting a graph into a tree

### PR DESCRIPTION
**Pull request by admons**
- [X] Ready for review
- [X] Follows [CONTRIBUTING](https://github.com/snyk/dep-graph/blob/master/.github/CONTRIBUTING.md) rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
It adds memoization in the `graphToDepTree`
This will result in reused objects in a tree instead, for example the following graph
```
A --> B --> D
  \-> C --> B
```
Instead of creating `B --> D` twice (as today), it will reuse the B subtree

### CODE REVIEW

**Comment by FauxFaux** [comment]
Does this just allocate forever?

**Comment by admons** [comment]
> Does this just allocate forever?

No, it's a scoped function, GC will clean `memoizationMap`.

```typescript
function a(b = new Map()) { return b; }
assert.notEqual(a(), a()); // 
```

**Comment by robcresswell** [comment]
This looks cool @admons! Any idea on the impact?

**Comment by admons** [comment]
@robcresswell thanks :)
1. This will create trees that today would get stuck and won't be generated (i.e. root -> 100,000 children -> all pointing to a specific node "B" -> 50,000 children)
For those trees, the problem will be on the serialisation phase (if needed)
2. In case that we have a code (in another library/service) that mutate a specific node in a tree, if this node is reused, there will be side affects

**Comment by robcresswell** [comment]
@admons Ah sorry I should've been more specific. I got the overall idea; I meant rather if you'd been able to benchmark anything. Just out of curiosity than anything else 🤓

**Comment by admons** [comment]
@robcresswell :)
I have a DepGraph of the [cockroachDB](https://github.com/cockroachdb/cockroach/blob/master/go.mod) and its an enormous graph [see here](https://drive.google.com/file/d/1ZoDlsc6F5qczTBWo_7bsyhFNVur11BzO/view?usp=sharing)
More than 15 minutes of 100% cpu to convert this depgraph still wasn't enought
After adding this change, it took 88ms

**Comment by robcresswell** [comment]
Niiice, thats what I wanted to hear 🤓

File: src/legacy/index.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import * as crypto from 'crypto';
 import * as types from '../core/types';
-import { DepGraphBuilder } from '../core/builder';
+import { DepGraphBuilder } from '..';
```
Comment (darscan) [inline src/legacy/index.ts:3]: Unneeded change. Please restore this line - the way it's importing right now creates more static coupling than before

File: src/legacy/index.ts [hunk 2]
```diff
@@ -286,7 +286,11 @@ async function buildSubtree(
   nodeId: string,
   eventLoopSpinner: EventLoopSpinner,
   maybeDeduplicationSet: Set<string> | null | false = null, // false = disabled; null = not in deduplication scope yet
```
Comment (darscan) [inline src/legacy/index.ts:288]: I know it shouldn't be part of this PR, but can we sneak this in?
```suggestion
  maybeDeduplicationSet: Set<string> | null | false = false, // false = disabled; null = not in deduplication scope yet
```
(default to false like we spoke about earlier)

**Review by darscan — APPROVED** [review]

**Comment by darscan** [comment]
🎉
